### PR TITLE
add ipcpath as argument

### PR DESCRIPTION
**Pull request by yann300**
https://github.com/ethereum/ethereum-console/issues/4

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -22,3 +27,23 @@ web3.eth.getBlockNumber(function(err, number) {
...
+
+function processArguments()
+{
+     var argIndex = 1;
+     for (var k = 2; k < process.argv.length; k++)
+     {
+     	var arg = process.argv[k];    
+        if (arg === "--ipcpath" && process.argv.length > k + 1)
+        {
+	    k++;
```
Comment (LianaHus) [inline main.js]: indent is wrong here

File: getIpcPath.js [hunk 2]
```diff
@@ -9,12 +9,12 @@ module.exports = function() {
     var path = process.env.HOME;
 
     if(process.platform === 'darwin')
-        path += '/Library/Ethereum/geth.ipc';
```
Comment (chriseth) [inline getIpcPath.js:12]: Please don't modify this file, it is ... erm ... copied from somewhere else.

**Comment by chriseth** [comment]
Sorry, I should have probably explained the story butter.
If you use `geth attach` with a different path, you just specify the location of the file itself, and not only some parent directory. I think it should work as follows:
`ethconsole /some/dir` -> attach to `/some/dir/geth.ipc`
`ethconsole /some/dir/geth.ipc` -> attach to `/some/dir/geth.ipc`
`ethconsole ipc:///some/dir/geth.ipc` -> attach to `/some/dir/geth.ipc`
`ethconsole ipc:///some/dir/` -> attach to `/some/dir/geth.ipc`

Furthermore, I think there are two main arguments to `ethconsole`:
1. the ipc path
2. some script to execute

I think both of them can just be given directly as arguments because you can distinguish them from their file type, so in this story, we do not need to parse any `--ipcpath` options, but just assume that any argument is the path to the ipc socked.

**Comment by LianaHus** [comment]
sorry you have tabs instead of 4 spaces

**Comment by chriseth** [comment]
Is there a way to check whether the given argument is a (1) regular file, (2) directory or (3) socket and react on it accordingly: 1: execute as script, 2: connect to geth.ipc inside that directory, 3: connect to the socket. Plus it should also work on windows :-)

**Comment by winsvega** [comment]
https://github.com/ethereum/ethereum-console/pull/6

how 2) some script to execute 
will look like?

File: main.js [hunk 3]
```diff
@@ -1,24 +1,90 @@
...
+var fs = require('fs')
+var vm = require('vm');
 var ipcpath = require('./getIpcPath.js');
 
+var ipcPath = ipcpath();
+var jsScript;
+
+console.log("")
+console.log("------------------")
+console.log("Run by default in interactive mode. When called in script mode, process.exit() should be called in your script to exit the nodejs app.")
```
Comment (chriseth) [inline main.js]: Could we move that to some `--help` description?

File: main.js [hunk 4]
```diff
@@ -1,24 +1,90 @@
...
+console.log("Arguments:")
+console.log("- a path which target a JavaScript file to execute (.js extension).")
```
Comment (chriseth) [inline main.js]: `path to a JavaScript file`?

File: main.js [hunk 5]
```diff
@@ -1,24 +1,90 @@
...
+	for (var k = 2; k < process.argv.length; k++)
+	{
+		var arg = process.argv[k];
+		if (arg.endsWith('.js'))
+			jsScript = arg;
+		else
+		{
+			ipcPath = arg;
+			if (ipcPath.startsWith("ipc:"))
+				ipcPath = ipcPath.substring(4)
```
Comment (chriseth) [inline main.js]: semicolon

File: main.js [hunk 6]
```diff
@@ -1,24 +1,90 @@
...
+		}
+	}
+	return true;
+}
+
+function executeScript()
+{
+	console.log("Executing " + jsScript + " ...");
+	fs.readFile(jsScript, 'utf8', function (err, data) {
```
Comment (chriseth) [inline main.js]: `{` to the next line

File: main.js [hunk 7]
```diff
@@ -1,24 +1,105 @@
...
+		{
+			var script = new vm.Script(data);
+			script.runInThisContext();
+		}
+	});
+}
+
+function logHelp()
+{
+	console.log("Run by default in interactive mode. When called in script mode, process.exit() should be called in your script to exit the nodejs app.")
```
Comment (chriseth) [inline main.js]: We should probably also tell the people what this is. Please also avoid long lines. Perhaps like that:

```
Usage: ethconsole [javascript file] [ipc socket]
Connects to an ethereum node via ipc in order to control it remotely
through global variable web3 (web3.admin is also present).
If no arguments are given, connects to the ipc socket at <INSERT STANDARD SOCKET HERE>
and drops into interactive mode.
Arguments:
  <ipc socket path>  connect to the given ipc socket (use ipc://<path> if it does not end with .ipc)
  <javascript file>    execute the given javascript file that has to end in .js non-interactively.
                           The script has to call process.exit() in order to terminate the console.
```

File: main.js [hunk 8]
```diff
@@ -98,8 +98,14 @@ function executeScript()
 
 function logHelp()
 {
-	console.log("Run by default in interactive mode. When called in script mode, process.exit() should be called in your script to exit the nodejs app.")
-	console.log("Arguments:")
-	console.log("- a path to a JavaScript file to execute (.js extension).")
-	console.log("- a path to an ipc path.")
+	var help = "\nUsage: ethconsole [javascript file] [ipc socket]\n" +
```
Comment (chriseth) [inline main.js]: Why do you need `\n` at the beginning?

File: main.js [hunk 9]
```diff
@@ -98,8 +98,14 @@ function executeScript()
...
+	"Connects to an ethereum node via ipc in order to control it remotely\n" +
+	"through global variable web3 (web3.admin is also present).\n" +
+	"If no arguments are given, connects to the ipc socket at " + ipcpath() + "\n" +
+	"and drops into interactive mode.\n" +
+	"Arguments:\n" +
+	"  <ipc socket path>	connect to the given ipc socket (use ipc://<path> if it does not end with .ipc)\n" +
+	"  <javascript file>	execute the given javascript file that has to end in .js non-interactively.\n" +
+	"                           The script has to call process.exit() in order to terminate the console.\n"
```
Comment (chriseth) [inline main.js]: Remove four spaces in the string to align the lines?

**Pull request merged**